log4j-api/src/main/java/org/apache/logging/log4j/util/ServiceLoaderUtil.java

```diff
@@ -110,28 +135,64 @@ private ServiceLoaderUtil() {
      * @param logger      An action to perform for each broken service.
      * @param services    A list to add the services.
      */
-    private static <T> void addServices(final Class<T> serviceType, final ClassLoader classLoader,
-            final Consumer<Throwable> logger, final Collection<T> services) {
-        final Iterator<T> iterator = ServiceLoader.load(serviceType, classLoader).iterator();
-        while (iterator.hasNext()) {
-            try {
-                final T service = iterator.next();
-                if (classLoader.equals(service.getClass().getClassLoader())) {
-                    services.add(service);
-                }
-            } catch (ServiceConfigurationError e) {
-                if (logger != null) {
-                    logger.accept(e);
-                }
-            }
-        }
+    private static <T> Stream<T> loadClassloaderServices(final Class<T> serviceType, final ClassLoader classLoader,
+            final boolean verbose) {
+        return StreamSupport.stream(new ServiceLoaderSpliterator<T>(serviceType, classLoader, verbose), false);
     }
 
-    private static <T> void addOsgiServices(final Class<T> serviceType, final Collection<T> services) {
+    private static <T> Stream<? extends T> loadOsgiServices(final Class<T> serviceType) {
         final Iterable<? extends T> iterable = org.glassfish.hk2.osgiresourcelocator.ServiceLoader
                 .lookupProviderInstances(serviceType);
-        if (iterable != null) {
-            iterable.forEach(services::add);
+        return iterable != null ? StreamSupport.stream(iterable.spliterator(), false) : Stream.empty();
+    }
+
+    private static class ServiceLoaderSpliterator<S> implements Spliterator<S> {
```

Comment:
jdk9+ introduced similar functionality with https://docs.oracle.com/javase/9/docs/api/java/util/ServiceLoader.html#stream--

The upside is that `Provider.get()` (instantiating a default instance) can fail (classpath/linker issues) without breaking the entire call.
Probably not worth adding jdk11 mr-jar code to handle, but we should add a comment describing that the feature should be used when jdk11 is the minimum supported version

Reply:
I'll add `ServiceLoader#stream()` to the `ServiceLoaderUtil` in master, when this version reaches its final form.

Reply:
I used the `stream()` method in the Java 9 specific class, so we don't forget.